### PR TITLE
LOOP-350 low reservoir alerts

### PR DESCRIPTION
**Pull request by ps2**
Adds implementation of reservoir behavior and persistence to MockPodCommManager, in service of supporting low reservoir alerts.

### CODE REVIEW

File: DashKit/Mocks/MockPodCommManager.swift [hunk 1]
```diff
@@ -74,16 +78,64 @@ public class MockPodCommManager: PodCommManagerProtocol {
...
+            }
+            
+            DispatchQueue.main.asyncAfter(deadline: .now() + 11.5) {
+                // Start out with 100U
+                self.podStatus = MockPodStatus(activationDate: Date(), podState: .uidSet, programStatus: ProgramStatus(rawValue: 0), activeAlerts: PodAlerts(rawValue: 128), isOcclusionAlertActive: false, bolusUnitsRemaining: 0, initialInsulinAmount: 100)
+                eventListener(.event(.podStatus(self.podStatus!)))
+            }
+            
+            DispatchQueue.main.asyncAfter(deadline: .now() + 12) {
+                eventListener(.event(.programmingLumpOfCoal))
```
Comment (nhamming) [inline DashKit/Mocks/MockPodCommManager.swift:108]: :) bad Christmas?

File: DashKit/Mocks/MockPodCommManager.swift [hunk 2]
```diff
@@ -74,16 +78,64 @@ public class MockPodCommManager: PodCommManagerProtocol {
...
+                self.podStatus!.insulinDelivered = 1.40
+                eventListener(.event(.podStatus(self.podStatus!)))
+                eventListener(.event(.programmingPodExpireAlert))
+            }
+            
+            DispatchQueue.main.asyncAfter(deadline: .now() + 31) {
+                eventListener(.event(.podStatus(self.podStatus!)))
+            }
+
+            DispatchQueue.main.asyncAfter(deadline: .now() + 32) {
```
Comment (nhamming) [inline DashKit/Mocks/MockPodCommManager.swift:140]: Should this be even faster? 32 seconds is a long time to wait if you are testing.

File: DashKit/Mocks/MockPodCommManager.swift [hunk 3]
```diff
@@ -95,36 +147,78 @@ public class MockPodCommManager: PodCommManagerProtocol {
...
-                    programStatus: .basalRunning,
-                    activeAlerts: PodAlerts([]),
-                    isOcclusionAlertActive: false,
-                    bolusUnitsRemaining: 0,
-                    totalUnitsDelivered: 38,
-                    reservoirUnitsRemaining: 1023,
-                    timeElapsedSinceActivation: 2,
-                    activationTime: activation)
-                eventListener(.event(.podStatus(self.podStatus)))
+            DispatchQueue.main.asyncAfter(deadline: .now() + 16) {
```
Comment (nhamming) [inline DashKit/Mocks/MockPodCommManager.swift:202]: Similar here, should this be faster?

File: DashKit/Mocks/MockPodCommManager.swift [hunk 4]
```diff
@@ -205,26 +327,46 @@ public class MockPodCommManager: PodCommManagerProtocol {
         } else {
             switch programType {
             case .bolus:
+                self.podStatus?.cancelBolus()
```
Comment (nhamming) [inline DashKit/Mocks/MockPodCommManager.swift]: since you assign `podStatus` to `self.podStatus`, why not just use `podStatus` here and following?

File: DashKit/Mocks/MockPodCommManager.swift [hunk 5]
```diff
@@ -175,28 +277,48 @@ public class MockPodCommManager: PodCommManagerProtocol {
     }
 
     public func sendProgram(programType: ProgramType, beepOption: BeepOption?, completion: @escaping (PodCommResult<PodStatus>) -> ()) {
+        self.podStatus?.updateDelivery()
+
+        guard var _ = podStatus else {
+            completion(.failure(.podIsNotActive))
+            return
+        }
```
Comment (nhamming) [inline DashKit/Mocks/MockPodCommManager.swift]: you have a pattern in `public func stopProgram(programType: StopProgramType, completion: @escaping (PodCommResult<PodStatus>) -> ()) {` for checking and assigning `podStatus`. Should that be repeated here to get rid of the force unwrapping?

File: DashKit/Mocks/MockPodCommManager.swift [hunk 6]
```diff
@@ -205,26 +327,46 @@ public class MockPodCommManager: PodCommManagerProtocol {
...
 
     public func silenceAlerts(alert: PodAlerts, completion: @escaping (PodCommResult<PodStatus>) -> ()) {
+        guard let podStatus = podStatus else {
+            completion(.failure(.podIsNotActive))
+            return
+        }
+        self.podStatus!.activeAlerts = podStatus.activeAlerts.subtracting(alert)
         silencedAlerts.append(alert)
-        completion(.success(podStatus))
+        completion(.success(self.podStatus!))
```
Comment (nhamming) [inline DashKit/Mocks/MockPodCommManager.swift]: Why is this needed? If it is really needed, would line 370, 392, etc. also need this?

File: DashKit/Mocks/MockPodStatus.swift [hunk 7]
```diff
@@ -22,30 +23,245 @@ public struct MockPodStatus: PodStatus {
...
+                basalProgram: BasalProgram? = nil)
+    {
+        self.activationDate = activationDate
+        self.podState = podState
+        self.programStatus = programStatus
+        self.activeAlerts = activeAlerts
+        self.isOcclusionAlertActive = isOcclusionAlertActive
+        self.bolusUnitsRemaining = bolusUnitsRemaining
+        self.initialInsulinAmount = initialInsulinAmount
+        self.insulinDelivered = 0
```
Comment (nhamming) [inline DashKit/Mocks/MockPodStatus.swift]: should this be `self.insulinDelivered = insulinDelivered`?

File: DashKitUI/Mocks/MockDashSettingsViewModel.swift [hunk 8]
```diff
@@ -88,15 +88,15 @@ class MockDashSettingsViewModel: DashSettingsViewModelProtocol {
...
+//    static func noPod() -> MockDashSettingsViewModel {
+//        return MockDashSettingsViewModel()
+//    }
+//    
+//    static func livePod() -> MockDashSettingsViewModel {
+//        let model = MockDashSettingsViewModel()
+//        model.basalDeliveryState = .active(Date())
+//        model.lifeState = .timeRemaining(.days(2.5))
+//        return model
+//    }
```
Comment (nhamming) [inline DashKitUI/Mocks/MockDashSettingsViewModel.swift]: Is this needed?

**Review by nhamming — APPROVED** [review]
No blocking comments.

File: DashKit/Mocks/MockPodCommManager.swift (hunk 1 above)
Comment (ps2) [reply DashKit/Mocks/MockPodCommManager.swift:108]: Mommy and daddy are mad. Boy, are they mad!

File: DashKit/Mocks/MockPodCommManager.swift (hunk 2 above)
Comment (ps2) [reply DashKit/Mocks/MockPodCommManager.swift:140]: Sure. The goal of a lot of this area of changes was to make it more true to pod behavior, but I don't think anything major is impacted by making this shorter; the progress bar will go away "early", in the user's view, but we can explain that as just a feature of the demo driver.

Comment (ps2) [reply DashKit/Mocks/MockPodCommManager.swift:140]: On second thought, looking at the code; I'd have to bump a bunch of values down; I think it'd be best to leave this for now, and keep it fairly close to actual pod behavior.

File: DashKit/Mocks/MockPodCommManager.swift (hunk 4 above)
Comment (ps2) [reply DashKit/Mocks/MockPodCommManager.swift]: Oh, there's bugs here. Thanks for pointing it out.

File: DashKit/Mocks/MockPodCommManager.swift (hunk 6 above)
Comment (ps2) [reply DashKit/Mocks/MockPodCommManager.swift]: I'll change this to the pattern of modifying a local copy and then assigning back.

File: DashKitUI/Mocks/MockDashSettingsViewModel.swift (hunk 8 above)
Comment (ps2) [reply DashKitUI/Mocks/MockDashSettingsViewModel.swift]: Nope
